rust/zypp-agama/zypp-agama-sys/c-layer/callbacks.cxx

```diff
@@ -390,39 +389,158 @@ struct KeyRingReport
+    switch (response) {
+    case PROBLEM_RETRY:
+      return zypp::target::PatchScriptReport::RETRY;
+    case PROBLEM_ABORT:
+      return zypp::target::PatchScriptReport::ABORT;
+    case PROBLEM_IGNORE:
+      return zypp::target::PatchScriptReport::IGNORE;
+    }
+    // fallback that should not happen
+    return zypp::target::PatchScriptReport::ABORT;
+  }
+};
+
+static PatchScriptReport patch_script_report;
+
+struct InstallResolvableReport
+    : public zypp::callback::ReceiveReport<
+          zypp::target::rpm::InstallResolvableReport> {
+  struct InstallCallbacks *callbacks;
+
+  InstallResolvableReport() { callbacks = NULL; }
+
+  void set_callbacks(InstallCallbacks *callbacks_) { callbacks = callbacks_; }
+
+  void start(zypp::Resolvable::constPtr resolvable) override {
+    if (callbacks == NULL || callbacks->package_start == NULL) {
+      return;
+    }
+    callbacks->package_start(resolvable->name().c_str(),
+                             callbacks->package_start_data);
+  }
+
+  Action problem(
+      zypp::Resolvable::constPtr resolvable,
+      zypp::target::rpm::InstallResolvableReport::Error error,
+      const std::string &description,
+      // note: the RpmLevel argument is not used anymore, ignore it
+      zypp::target::rpm::InstallResolvableReport::RpmLevel _level) override {
+    if (callbacks == NULL || callbacks->package_problem == NULL) {
+      return zypp::target::rpm::InstallResolvableReport::problem(
+          resolvable, error, description, _level);
+    }
+    PROBLEM_RESPONSE response = callbacks->package_problem(
+        resolvable->name().c_str(), convert_error(error), description.c_str(),
+        callbacks->package_problem_data);
+    return convert_action(response);
+  }
+
+  void finish(
+      zypp::Resolvable::constPtr resolvable, Error error,
+      const std::string &install_info,
+      zypp::target::rpm::InstallResolvableReport::RpmLevel /*level*/) override {
+    if (callbacks == NULL || callbacks->package_finish == NULL) {
+      return;
+    }
+    callbacks->package_finish(resolvable->name().c_str(),
+                              callbacks->package_finish_data);
+  }
+
+private:
+  inline ZyppInstallPackageError
+  convert_error(zypp::target::rpm::InstallResolvableReport::Error error) {
+    switch (error) {
+    case zypp::target::rpm::InstallResolvableReport::Error::NO_ERROR:
+      return ZyppInstallPackageError::PI_NO_ERROR;
+    case zypp::target::rpm::InstallResolvableReport::Error::NOT_FOUND:
+      return ZyppInstallPackageError::PI_NOT_FOUND;
+    case zypp::target::rpm::InstallResolvableReport::Error::IO:
+      return ZyppInstallPackageError::PI_IO;
+    case zypp::target::rpm::InstallResolvableReport::Error::INVALID:
+      return ZyppInstallPackageError::PI_INVALID;
+    }
+    return ZyppInstallPackageError::PI_NO_ERROR;
+  }
+
+  inline zypp::target::rpm::InstallResolvableReport::Action
+  convert_action(PROBLEM_RESPONSE response) {
```

Comment:
Naming: I would expect `convert_response` since that is the argument type.

(Actually I would like to see Rust-like function names, like `to_foo`, `as_foo`  for standalone functions or `from_foo`  for methods)